### PR TITLE
[WIP][Speculative Decoding] Enable batch size > 1

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR enables speculative decoding when batch size is > 1.

This PR has been tested to work correctly for Whisper. You can try it as follows:

```py
from transformers import pipeline, AutoModelForCausalLM, AutoModelForSpeechSeq2Seq, AutoProcessor
import torch
from datasets import load_dataset

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

assistant_model_id = "distil-whisper/distil-large-v2"

assistant_model = AutoModelForCausalLM.from_pretrained(
    assistant_model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
assistant_model.to(device)

model_id = "openai/whisper-large-v2"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    generate_kwargs={"assistant_model": assistant_model},
    torch_dtype=torch_dtype,
    chunk_length_s=15,
    batch_size=4,
    device=device,
)

dataset = load_dataset("distil-whisper/librispeech_long", "default", split="validation")
sample = dataset[0]["audio"]

result = pipe(sample)
print(result["text"])
```

## TODOS
- [x] Test that it works correctly for Whisper
- [ ] Test for Llama
- [ ] Write nice tests
- [ ] Clean code
- [ ] Write docs

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1610,8 +1610,8 @@ def generate(
                     "num_return_sequences has to be 1 when doing assisted generate, "
                     f"but is {generation_config.num_return_sequences}."
                 )
-            if batch_size > 1:
-                raise ValueError("assisted generate is only supported for batch_size = 1")
+            # if batch_size > 1:
+            #    raise ValueError("assisted generate is only supported for batch_size = 1")
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:1614]:
```suggestion
```

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -4555,40 +4602,58 @@ def assisted_decoding(
...
-            n_matches = min(n_matches, max_len - cur_len - 1)
+            n_matches -= last_assistant_token_is_eos.int() * (n_matches == candidate_length).int()
+            # make sure than already finished sequences always match until longest "still active" sequence
+            n_matches = torch.clamp(n_matches, max=max_len - cur_len - 1)
+            # make sure that finished sentences cannot slow down valid tokens
+            n_matches = unfinished_sequences * n_matches + (1 - unfinished_sequences) * (unfinished_sequences * n_matches).max()
 
             # 5.2. Get the valid continuation, after the matching tokens
-            valid_tokens = selected_tokens[:, : n_matches + 1]
+            valid_tokens = selected_tokens[:, : n_matches.max() + 1]
```
Comment (MahmoudAshraf97) [inline src/transformers/generation/utils.py:4620]: I have a question about a part that I don't understand, given that the results are identical with and without batched spec. decoding

this assumes that all the sequences will have the same number of correct matches, when we choose the accepted tokens number based on the maximum correct matches, we also accept wrong tokens from sequences that had lower number of correct matches than the maximum, the solution would be to choose the minimum which guarantees that all the accepted tokens are correct
```suggestion
            valid_tokens = selected_tokens[:, : n_matches.min() + 1]
```
that poses a risk that if a single sequence had zero correct matches, this whole spec. decoding cycle will be discarded for all other sequences even if they have correct matches.

is it  possible to use padding to accept different number of candidate tokens per batch element as opposed to accepting the minimum/maximum common number of correct suggested tokens across the batch, this should at least keep the speedups of spec decoding that were observed at small batch sizes (if not even faster at larger sizer)

**Comment by preminstrel** [comment]
Hi, good work! I am curious about how to deal with the wrong kv cache token rolling back for draft model, or you just mask them?

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (learning-chip) [reply src/transformers/generation/utils.py:4620]: > is it possible to use padding to accept different number of candidate tokens per batch element as opposed to accepting the minimum/maximum common number of correct suggested tokens across the batch

This is possible if the padding lengths are adjusted every iteration.

**Comment by Hubert-Bonisseur** [comment]
@patrickvonplaten Is there any plan to merge this PR sometimes or is it canceled ?

**Comment by amyeroberts** [comment]
@Hubert-Bonisseur The PR has closed due to inactivity. I'm not sure of the current status of batch > 1 for speculative decoding, but I'm sure @gante will know :)

**Comment by Hubert-Bonisseur** [comment]
It is still not supported : 
`ValueError: assisted generate is only supported for batch_size = 1`

But I see that Patrick has left HF since writing this PR, which probably explains why it stalled.
If there is still interest in this PR, I'm willing to help with finish the work of Patrick. It seems like it is very close to be ready and I'm vert interested in this feature.

**Comment by Hubert-Bonisseur** [comment]
Maybe your whisper expert @sanchit-gandhi also knows the status of this feature ?

**Comment by sanchit-gandhi** [comment]
Hey @Hubert-Bonisseur - handing this one over to @kamilakesbi who will take over the integration!

**Comment by Hubert-Bonisseur** [comment]
Hey @kamilakesbi, just checking in to see if there are any updates on this. Thanks!

**Pull request closed without merging**